### DISCUSSION TITLE
Implementing a new type of camera in crappy

### DISCUSSION THREAD

**Discussion by Hyo-u** [LaboratoireMecaniqueLille/crappy#34] [Q&A]
Hi there,

### Background:
Elbub supervisor here, first thing first, thank you for building Crappy, it clearly helps (at least our) small company that want to devellop their R&D. 
The program built by Elbub on my specification works well (there is still sometime some error when reloading but nothing truly problematic).
But there are still some part to smooth up/implement/clean which will probably lead to another internship next year.
Disclamer : I don't usually code so take everything I say with a grain of salt (And even if I learned some python to follow Elbub internship, it is still very new to me).

### What I want :
I have two cameras, I want them to take pictures (with a lower frequency than others inputs) 
Store the time of when the picture is taken (to match with load and displacement) and their numbers in a file.

### What has been done:
Tried : Take a picture with python (file : Test_camVimba)
	Result : Work
Tried : use the camera with the open_cv module in crappy
	Result : doesn't work
	Reason possible : I think because open_cv doesn't work alone for accessing the camera
Tried : Create a custom camera class in crappy (file : Vimba.py/Camera_Vimba_0.1.py), add it to the crappy/camera/__init__.py and try to launch a camera block: Test_CamVimba_crappy.py)
	Result : doesn't work
	Reason possible : There should be some or a lot of mistakes in my code.

### Question :
What I want to do doesn't seem that difficult on paper, but I can't wrap my head around how to add a camera to crappy. 
Is there something obvious I doesn't see and have you any advice?

### Info : 
My camera : Alvium 1800 C-1620
Connector : USB3
OS : Windows 11
Python 3.10
Crappy 1.5.9
The Python librairy for my camera :
VimbaPython 1.2.1
Opencv 4.7.0.68
My files : [Camera.zip](https://github.com/LaboratoireMecaniqueLille/crappy/files/10871921/Camera.zip)

**Comment by WeisLeDocto** [comment]
Hello @Hyo-u,

Glad to hear that we can be of some help in your work !

For interfacing your camera with Crappy, I strongly suggest that you define a class in your test script (or in an other file imported in your test script), as shown [here](https://github.com/LaboratoireMecaniqueLille/crappy/blob/0485bfc77b3196a07fcbe7e335178bd37b38f1e7/Examples/custom_camera.py). In this case, there's no need to change anything in the source code of Crappy, including the `__init__.py` files. Hopefully the [dedicated section in the documentation](https://github.com/LaboratoireMecaniqueLille/crappy/blob/0485bfc77b3196a07fcbe7e335178bd37b38f1e7/Examples/custom_camera.py) can be of some help to you.

I took a quick look at the `vimba` library that you want to use, it sems totally feasible to drive your camera from Crappy. If you're lost, I suggest you to modify your `Test_camVimba.py` script so that it doesn't include any context manager (i.e. `with` statement, hopefully Vimba allows it). Then you'll get a basic sequential script, and you'll be able to split the lines between the `open`, `get_image` and `close` methods of the Camera object.  At first, avoid adding settings to your camera and keep your code as basic as possible.

Also, I'm not sure why you tried to use `cv2.VideoCapture` in the scripts that you sent, but if you can't read your camera with the built-in `Camera_opencv` of Crappy, then you'll have to rely on `vimba` only for grabbing the frame (like in your Test_camVimba.py example).

I don't have time now to dig in the `vimba` library, but I'll gladly help if you need further help with Crappy.
Weis

P.S. : Instead of just saying that your script doesn't work, please provide the Traceback (the console output) containing the error that you get. It will help understanding what's wrong with your code.

**Reply by Hyo-u** [reply]
Thank you for your detailed answer @WeisLeDocto ,

That is sort of what I have tried to do with my vimba.py. But I get error like this :
```
106, in __init__
    assert self.camera_name in camera_list or self.input_label,\
AssertionError: My_cam camera does not exist!
```
Ok good to know, modifying the `__init__.py` did cause some problem.

In the manual of Vimba Python, from what I understand it is specified to always use a `with `statement :

> The Camera class implements a context manager. On entering the Camera's context, all camera features are detected and can be accessed only within the with statement (see pdf) .
```
from vimba import *
with Vimba.get_instance () as vimba:
cams = vimba.get_all_cameras ()
```
When I try to bypass it, I got this error :
`RuntimeError: Called 'Vimba.__Impl.get_all_cameras()' outside of 'with' - statement scope.`

From what I have read, Vimba use a **GenICam transport layer (GenTL)** to communicate, I will dig if I can find something useful without a `with` statement. If not I dunno what I can do. 

PS : Yes it would have been more helpful, but I wasn't on the same PC as the one on which I tested my script. Will definitely do it when I come close to something.

[Vimba_Python_Manual (1).pdf](https://github.com/LaboratoireMecaniqueLille/crappy/files/10897403/Vimba_Python_Manual.1.pdf)

**Reply by WeisLeDocto** [reply]
> That is sort of what I have tried to do with my vimba.py. But I get error like this :

In the `vimba.py` example you define a class called `'Camera_Vimba'`, but when you then instantiate the Camera Block you pass the name `'My_cam'` as an argument ! Hence the error that you get. You have to modify either of these names so that they match. Also, the first letter of the class must be capital and the others lowercase (will be changed in the next release).


Regarding the context manager issue, I get it that every call to Vimba's API has to be done in the context of a Vimba instance, and calls to Camera methods have to be done in the context of a Camera object. A Vimba instance is simply what is returned by `Vimba.get_instance()`, and is a regular Python object (see manual page 14). In a similar fashion, Vimba Camera objects are regular Python objects, so you can maybe try something like : 
```
from vimba import *
vimba = Vimba.get_instance()
with vimba:
    cam = vimba.get_all_cameras()[0]
with vimba:
    with cam:
        frame = cam.get_frame()
```
Notice the two consecutive contexts of the same instance. If that succeeds, then you'll be able to split the calls between the several methods of a `crappy.camera.Camera` class, with each method its own context manager(s). If not, there are other possible workarounds but it might become less elegant.

Weis

**Reply by Hyo-u** [reply]
It is starting to work thanks to you (using your advices). 
The camera is detected, Crappy launch now, tkinter open, and it loop the `get_image` function. I just see a black picture for now (I think, I still don't understand much but I'm getting there). I will probably be back with a lot more questions later, when I have more time to work on it.

I got an error that needed me to change a line in a crappy file : `cameraConfig.py `
      from : `fx = np.arange(0, out_size[1], out_size[1] / l0, dtype=np.float)`
      to : `fx = np.arange(0, out_size[1], out_size[1] / l0, dtype=float)`
With this error : AttributeError: module 'numpy' has no attribute 'float'.
`np.float` was a deprecated alias for the builtin `float`.

My code as of today :
```
import crappy
import numpy as np
from vimba import *
from time import time
import os

class Camera_vimba(crappy.camera.Camera):

    def __init__(self, resolution=(3040, 5312)):
        super().__init__()
        self.resolution = resolution
        self.frame = None
        self.name = "camera_vimba"
        self.add_setting('level', self._get_lvl, self._set_lvl, limits=(0, 255))

    def _get_lvl(self):
        return self.frame[0, 0]

    def _set_lvl(self, val):
        self.frame = np.ones(self.resolution, dtype=np.uint8) * val
        return val

    def open(self, **kwargs):
        global cam
        global vimba
        vimba = Vimba.get_instance()
        with vimba:
            cam = vimba.get_all_cameras()[0] 
            print(cam)
        self.frame = np.zeros(self.resolution, dtype=np.uint8)
        self.set_all(**kwargs)

    def get_image(self):
        with vimba:
            with cam:
                frame = cam.get_frame()
                print('got Frame')
        return time(), self.frame

    def close(self):
        print("Closing")
        del self.frame

if __name__ == "__main__":
    camera = crappy.blocks.Camera(camera='Camera_vimba', verbose=True, display_images=True, displayer_framerate=1)
    crappy.start()

```

**Reply by WeisLeDocto** [reply]
Good to hear you're making progress !

The deprecation error with `numpy.float` was normally fixed with commit 24715b2.

Took a quick look at your code : you could get rid of the `resolution` attribute and argument, as well as of the `level` setting and its getter and setter. The `self.frame` attribute is also not necessary. Also, instead of defining `global` variables, you should make them attributes of your class (using `self.<name of attribute>`). 

The reason why you're getting a black image is because you acquire `frame` but return `self.frame`, these are not the same objects ! And one last thing, you'll probably have to use smth like `np_frame = frame.as_numpy_ndarray()` for converting the frame to the right format for Crappy to handle it.

Weis

**Comment by WeisLeDocto** [comment]
Closing discussion due to prolonged inactivity
